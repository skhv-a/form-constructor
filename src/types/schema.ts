export type NumberSchema = Partial<{
  required: boolean;
  min: number;
  max: number;
}>;

export type StringSchema = Partial<{
  required: boolean;
  min: number;
  max: number;
  pattern: string;
}>;

export type SelectSchema = Partial<{
  required: boolean;
}>;

export type PhoneSchema = Partial<{
  required: boolean;
}>;

export type FieldType =
  | "string"
  | "number"
  | "password"
  | "checkbox"
  | "phone"
  | "select";

export type FieldOption = {
  key: string;
  value: string;
};

export type SchemaValidation =
  | NumberSchema
  | StringSchema
  | SelectSchema
  | PhoneSchema;

export type SchemaField = {
  key: string;
  label: string;
  type: FieldType;
  validation: SchemaValidation;
  options?: FieldOption[];
};

export type Schema = {
  name: string;
  fields: SchemaField[];
};
