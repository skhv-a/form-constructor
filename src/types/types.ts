export type ValueOf<T> = T[keyof T];

export type TypeNestedValues<Values, T> = {
  [K in keyof Values]: Values[K] extends any[]
    ? Values[K][number] extends object
      ? TypeNestedValues<Values[K][number], T>[]
      : T[]
    : Values[K] extends object
    ? TypeNestedValues<Values[K], T>
    : T;
};
