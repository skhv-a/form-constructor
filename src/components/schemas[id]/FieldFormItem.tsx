import Checkbox from "components/Checkbox/Checkbox";
import FormItem from "components/Form/Item";
import Input from "components/Input/Input";
import NumberInput from "components/NumberInput";
import PasswordInput from "components/PasswordInput/PasswordInput";
import Select from "components/Select/Select";
import { SchemaField } from "types/schema";

type Props = { field: SchemaField };

const FieldFormItem = ({ field }: Props) => {
  switch (field.type) {
    case "string":
      return (
        <FormItem
          name={field.key}
          label={field.label}
          schema={{ type: "string", ...field.validation }}
        >
          <Input placeholder={placeholderFromLabel(field.label)} />
        </FormItem>
      );
    case "password":
      return (
        <FormItem
          name={field.key}
          label={field.label}
          schema={{ type: "string", ...field.validation }}
        >
          <PasswordInput placeholder={placeholderFromLabel(field.label)} />
        </FormItem>
      );
    case "phone":
      return (
        <FormItem
          name={field.key}
          label={field.label}
          schema={{ type: "phone", ...field.validation }}
        >
          <Input placeholder={placeholderFromLabel(field.label)} />
        </FormItem>
      );
    case "number":
      return (
        <FormItem
          name={field.key}
          label={field.label}
          schema={{ type: "number", ...field.validation }}
        >
          <NumberInput placeholder={placeholderFromLabel(field.label)} />
        </FormItem>
      );
    case "select":
      return (
        <FormItem
          name={field.key}
          label={field.label}
          schema={{ type: "select", ...field.validation }}
          valuePropName="selected"
          trigger="onSelect"
        >
          <Select
            placeholder={placeholderFromLabel(field.label)}
            options={field.options!.map((option) => ({
              label: option.value,
              value: option.key,
            }))}
          />
        </FormItem>
      );
    case "checkbox":
      return (
        <FormItem
          name={field.key}
          schema={{ type: "checkbox", ...field.validation }}
          valuePropName="checked"
          trigger="onCheck"
        >
          <Checkbox label={field.label} />
        </FormItem>
      );
    default:
      return null;
  }
};

function placeholderFromLabel(label: string): string {
  return `Введите ${label.toLowerCase()}`;
}

export default FieldFormItem;
