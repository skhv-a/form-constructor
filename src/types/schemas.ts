export type NumberSchema = Partial<{
  required: boolean;
  min: number;
  max: number;
}>;

export type StringSchema = Partial<{
  required: boolean;
  min: number;
  max: number;
  pattern: string;
}>;

type TypeWithSchema<T, S> = { type: T } & S;

export type Schema =
  | TypeWithSchema<"number", NumberSchema>
  | TypeWithSchema<"string", StringSchema>
  | TypeWithSchema<"select", { required?: boolean }>
  | TypeWithSchema<"phone", { required?: boolean }>;
