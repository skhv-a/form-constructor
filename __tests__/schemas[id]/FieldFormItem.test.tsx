import { shallow, ShallowWrapper } from "enzyme";
import FieldFormItem from "components/schemas[id]/FieldFormItem";
import { SchemaField } from "types/schema";

type Props = { field: SchemaField };

describe("FieldFormItem", () => {
  let wrapper: ShallowWrapper<Props>;

  const renderWrapper = (props: Props): void => {
    wrapper = shallow(<FieldFormItem {...props} />);
  };

  it("string", () => {
    renderWrapper({
      field: {
        key: "string",
        label: "строка",
        type: "string",
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("password", () => {
    renderWrapper({
      field: {
        key: "password",
        label: "пароль",
        type: "password",
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("phone", () => {
    renderWrapper({
      field: {
        key: "phone",
        label: "телефон",
        type: "phone",
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("number", () => {
    renderWrapper({
      field: {
        key: "number",
        label: "число",
        type: "number",
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("select", () => {
    renderWrapper({
      field: {
        key: "select",
        label: "значение из селекта",
        type: "select",
        options: [
          {
            key: "foo",
            value: "Первое значение",
          },
          {
            key: "bar",
            value: "Второе значение",
          },
        ],
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("checkbox", () => {
    renderWrapper({
      field: {
        key: "checkbox",
        label: "чекбокс",
        type: "checkbox",
        validation: {},
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it("validation passed", () => {
    renderWrapper({
      field: {
        key: "string",
        label: "строка",
        type: "string",
        validation: {
          required: true,
          min: 5,
          max: 10,
          pattern: "\\d+",
        },
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
