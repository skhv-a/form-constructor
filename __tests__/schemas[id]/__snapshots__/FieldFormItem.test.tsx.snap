// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldFormItem checkbox 1`] = `
<FormItem
  name="checkbox"
  schema={
    Object {
      "type": "checkbox",
    }
  }
  trigger="onCheck"
  valuePropName="checked"
>
  <Checkbox
    label="чекбокс"
  />
</FormItem>
`;

exports[`FieldFormItem number 1`] = `
<FormItem
  label="число"
  name="number"
  schema={
    Object {
      "type": "number",
    }
  }
>
  <NumberInput
    placeholder="Введите число"
  />
</FormItem>
`;

exports[`FieldFormItem password 1`] = `
<FormItem
  label="пароль"
  name="password"
  schema={
    Object {
      "type": "string",
    }
  }
>
  <PasswordInput
    placeholder="Введите пароль"
  />
</FormItem>
`;

exports[`FieldFormItem phone 1`] = `
<FormItem
  label="телефон"
  name="phone"
  schema={
    Object {
      "type": "phone",
    }
  }
>
  <ForwardRef
    placeholder="Введите телефон"
  />
</FormItem>
`;

exports[`FieldFormItem select 1`] = `
<FormItem
  label="значение из селекта"
  name="select"
  schema={
    Object {
      "type": "select",
    }
  }
  trigger="onSelect"
  valuePropName="selected"
>
  <Select
    options={
      Array [
        Object {
          "label": "Первое значение",
          "value": "foo",
        },
        Object {
          "label": "Второе значение",
          "value": "bar",
        },
      ]
    }
    placeholder="Введите значение из селекта"
  />
</FormItem>
`;

exports[`FieldFormItem string 1`] = `
<FormItem
  label="строка"
  name="string"
  schema={
    Object {
      "type": "string",
    }
  }
>
  <ForwardRef
    placeholder="Введите строка"
  />
</FormItem>
`;

exports[`FieldFormItem validation passed 1`] = `
<FormItem
  label="строка"
  name="string"
  schema={
    Object {
      "max": 10,
      "min": 5,
      "pattern": "\\\\d+",
      "required": true,
      "type": "string",
    }
  }
>
  <ForwardRef
    placeholder="Введите строка"
  />
</FormItem>
`;
